### ISSUE TITLE
C# use of enum type - should be optional

### ISSUE THREAD

**Issue by JensMeinecke** [dotnet/roslyn#8244]
Something that's been bugging me a long time. Now that C# 6 has come up with `using static` concept, I have decided to bring this up for discussion:

When dealing with enums we have constructs similar to the following:

``` C#
DialogResult result = DialogResult.Ok;
```

Now I fully understand why I have to put the first `DialogResult` in that declaration. But I don't understand why the enum type is required in the assignment part. If I omit it the compiler will complain, telling me that a `DialogResult` is expected. So if it know that we are dealing with `DialogResult` surely  it could look in the `DialogResult` type and allow

``` C#
DialogResult result = Ok;
```

This becomes particularly handy when dealing with `[Flag]` enums. Would your rather type all of this:

``` C#
this.Anchor = AnchorStyles.Bottom | AnchorStyles.Left | AnchorStyles.Right | AnchorStyles.Top;
```

or all of this:

``` C#
this.Anchor = Bottom | Left | Right | Top;
```

In the context of the assignment, it is pretty obvious what is meant. No need to guess.

Similarly with switch statement and comparisons:

``` C#
switch(form.DialogResult)
{ 
      case Ok:
           ////
           break;
      case Cancel:
           ////
           break;
}

if (form.DialogResult == Ok)
{
  /// do something
}
```

and, of course, return values:

``` C#
private DialogResult Result 
{      get { return Ok; } }
```

and function calls:

``` C#
private SetResult(DialogResult result);
....
SetResult(Ok);
```

In other words **in all context where an expression of the known type is expected**, you can infer the type.

If there is a conflict you have to resolve the ambiguity explicitly:

``` C#
class MyClass 
{
       int Left { get; set;}
       int Right {get; set;}

      void SetAnchorSide(int margin) {....}

      void SetAnchorSide(AnchorStyles side) { ...}

     void SomeCode()
     {
          // Ambiguous - could refer to either this.Left of AnchorStyles.Left
          SetAnchor(Left); 

         // Use either of the following two
         SetAnchor(AnchorStyles.Left);
         SetAnchor(this.Left);
    }
}

```

**Comment by alrz** [comment]
Dupe of #952
